demo/c/porcupine_demo_mic.c

```diff
@@ -9,23 +9,154 @@
+#else
+    return dlsym(handle, symbol);
+#endif
+}
+
+void close_dl(void *handle) {
+#if defined(WIN32) || defined(WIN64)
+    FreeLibrary((HMODULE) handle);
+#else
+    dlclose(handle);
+#endif
+}
+
+void print_dl_error(const char *message) {
+#if defined(WIN32) || defined(WIN64)
+    fprintf(stderr, "%s with code '%lu'.\n", message, GetLastError());
+#else
+    fprintf(stderr, "%s with '%s'.\n", message, dlerror());
+#endif
+}
+
 void interrupt_handler(int _) {
     (void) _;
     is_interrupted = true;
 }
 
+void data_callback_fixed(ma_device* device, void* output, const void* input, ma_uint32 frame_count){
+    pv_porcupine_data_t *pv_porcupine_data = (pv_porcupine_data_t *) device->pUserData;
+
+    int16_t *pcm = (int16_t *) input;
+
+    int32_t keyword_index = -1;
+    pv_status_t status = pv_porcupine_data->pv_porcupine_process_func(pv_porcupine_data->porcupine, pcm, &keyword_index);
+    if (status != PV_STATUS_SUCCESS) {
+        fprintf(stderr, "'pv_porcupine_process' failed with '%s'\n", pv_porcupine_data->pv_status_to_string_func(status));
+        exit(1);
+    }
+    if (keyword_index != -1) {
+        fprintf(stdout, "keyword detected\n");
+        fflush(stdout);
+    }
+
+    (void)output;
+    (void)frame_count;
+}
+
+void data_callback(ma_device* device, void* output, const void* input, ma_uint32 frame_count) {
+    pv_porcupine_data_t *pv_porcupine_data = (pv_porcupine_data_t *) device->pUserData;
+
+    frame_buffer_t *frame_buffer = &pv_porcupine_data->frame_buffer;
+
+    int16_t *buffer_ptr = &frame_buffer->buffer[frame_buffer->filled];
+    int16_t *input_ptr = (int16_t *) input;
+
+    int32_t processed_frames = 0;
+
+    while (processed_frames < frame_count) {
+        int32_t remaining_frames = (int32_t) frame_count - processed_frames;
+        int32_t available_frames = frame_buffer->max_size - frame_buffer->filled;
+
+        if (available_frames > 0) {
+            int32_t frames_to_read = (remaining_frames < available_frames) ? remaining_frames : available_frames;
+
+            memcpy(buffer_ptr, input_ptr, frames_to_read * sizeof(int16_t));
+            buffer_ptr += frames_to_read;
+            input_ptr += frames_to_read;
+
+            processed_frames += frames_to_read;
+            frame_buffer->filled += frames_to_read;
+        } else {
+            data_callback_fixed(device, output, frame_buffer->buffer, frame_count);
+
+            buffer_ptr = frame_buffer->buffer;
+            frame_buffer->filled = 0;
+        }
+    }
+}
+
 int main(int argc, char *argv[]) {
-    if (argc != 6) {
-        fprintf(stderr, "usage : %s library_path model_path keyword_path sensitivity input_audio_device\n", argv[0]);
+    ma_context context;
+    ma_device_info *capture_info;
+    ma_uint32 capture_count;
+    ma_result result;
+
+    result = ma_context_init(NULL, 0, NULL, &context);
+    if (result != MA_SUCCESS) {
+        fprintf(stderr, "failed to initialize context.\n");
```

Comment:
what does `context` mean here? go with more descriptive error message: `failed to initialize the input audio device list`? or something like that.